gatling-decoupled-response/src/main/scala/io/gatling/decoupled/state/PendingRequestsActor.scala

```diff
@@ -0,0 +1,225 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gatling.decoupled.state
+
+import java.time.Instant
+import java.util.UUID
+
+import akka.actor.{ Actor, ActorLogging, ActorRef, ActorRefFactory, Props, Timers }
+import io.gatling.commons.stats.{ KO, OK }
+import io.gatling.commons.util.Clock
+import io.gatling.core.action.Action
+import io.gatling.core.session.Session
+import io.gatling.core.stats.StatsEngine
+
+import scala.concurrent.duration.FiniteDuration
+
+object PendingRequestsActor {
+
+  def apply(actorRefFactory: ActorRefFactory, statsEngine: StatsEngine, clock: Clock, pendingTimeout: FiniteDuration): ActorRef = {
+    actorRefFactory.actorOf(
+      Props(new PendingRequestsActor(statsEngine, clock, pendingTimeout))
+    )
+  }
+
+  final case class ActorState(
+      waitingResponse: Map[UUID, RequestTriggered],
+      waitingTrigger: Map[UUID, DecoupledResponseReceived]
+  ) {
+    def withPendingResponse(response: DecoupledResponseReceived): ActorState = {
+      copy(waitingTrigger = waitingTrigger + (response.id -> response))
+    }
+
+    def withPendingTrigger(trigger: RequestTriggered): ActorState = {
+      copy(waitingResponse = waitingResponse + (trigger.id -> trigger))
+    }
+
+    def filterManagedId(id: UUID): ActorState = {
+      copy(waitingResponse = waitingResponse - id, waitingTrigger = waitingTrigger - id)
+    }
+
+  }
+
+  object ActorState {
+    val empty: ActorState = ActorState(Map.empty, Map.empty)
+  }
+
+  final case class ExecutionPhase(name: String, time: Instant)
+
+  sealed trait ActorMessage
+  final case class RequestTriggered(id: UUID, initialPhase: ExecutionPhase, session: Session, next: Action) extends ActorMessage
+  final case class DecoupledResponseReceived(id: UUID, executionPhases: Seq[ExecutionPhase]) extends ActorMessage
+  final case class WaitResponseTimeout(id: UUID) extends ActorMessage
+  final case class WaitTriggerTimeout(id: UUID) extends ActorMessage
+
+  val triggerPhaseName = "Trigger"
+  def genName(id: UUID, from: String, to: String): String = {
+    s"$id: $from -> $to"
+  }
+  val errorName = "Error"
+  def genErrorName(id: UUID): String = {
+    s"$id: $errorName"
+  }
+}
+
+class PendingRequestsActor(statsEngine: StatsEngine, clock: Clock, pendingTimeout: FiniteDuration) extends Actor with ActorLogging with Timers {
+  import PendingRequestsActor._
+
+  override def receive: Receive = receiveWithState(ActorState.empty)
+
+  private def receiveWithState(state: ActorState): Receive = {
+
+    case trigger @ RequestTriggered(id, _, _, _) if state.waitingResponse.contains(id) =>
+      log.debug("Duplicate trigger received: {}", trigger)
```

Comment:
Should not be a warning?

Reply:
when running load tests you risk killing everything if you start logging errors: logging is much heavier than one would expect. 
Where possible errors are captured by the statsEngine call in the `onWrongMessageReceived` code
The debug here is in case you see a lot of those errors in the final report, than you can run a small batch with debug turned on and see what is happening in detail

Reply:
hmm, I see.